core/src/main/scala/neotypes/Transaction.scala

```diff
@@ -1,120 +1,133 @@
     S.fToS(
-      F.async[S[T]] { cb =>
-        transaction.runAsync(query, convertParams(params))
-          .thenAccept { (x: StatementResultCursor) =>
-            cb {
-              Right(S.init[T](() => F.map(nextAsyncToF(x.nextAsync()))(Option(_))))
-            }
-          }
-          .exceptionally(exceptionally(ex =>
-            cb(Left(ex))
-          ))
+      F.async { cb =>
+        transaction
+          .runAsync(query, convertParams(params))
+          .accept { x: StatementResultCursor =>
+            cb(
+              Right(
+                S.init(
+                  () => nextAsyncToF(x.nextAsync())
+                )
+              )
+            )
+          }.recover { ex: Throwable => cb(Left(ex)) }
       }
     )
-  }
-
-  def commit(): F[Unit] = F.async(cb =>
-    transaction.commitAsync()
-      .thenAccept { _: Void => cb(Right(())) }
-      .exceptionally(exceptionally(ex => cb(Left(ex))))
-  )
 
-  def rollback(): F[Unit] = F.async(cb =>
-    transaction.rollbackAsync()
-      .thenAccept { _: Void => cb(Right(())) }
-      .exceptionally(exceptionally(ex => cb(Left(ex))))
-  )
+  def commit(): F[Unit] =
+    F.async{ cb =>
+      transaction
+        .commitAsync()
+        .accept { _: Void => cb(Right(())) }
+        .recover { ex: Throwable => cb(Left(ex)) }
+    }
 
-  private[this] def recordToSeq(record: Record): Seq[(String, Value)] = record.fields().asScala.map(p => p.key() -> p.value())
+  def rollback(): F[Unit] =
+    F.async { cb =>
+      transaction
+        .rollbackAsync()
+        .accept { _: Void => cb(Right(())) }
+        .recover { ex => cb(Left(ex)) }
+    }
 }
 
 object Transaction {
-  def convertParams(params: Map[String, Any]): util.Map[String, Object] = {
-    params.map {
-      case (k, v) => k -> toNeoType(v)
-    }.asJava.asInstanceOf[util.Map[String, Object]]
-  }
-
-  def toNeoType(value: Any): Any = value match {
-    case s: Seq[Any] => new util.ArrayList[Any](s.map(toNeoType).asJava)
-    case m: Map[Any, Any] => new util.HashMap[Any, Any](m.mapValues(toNeoType).asJava)
-    case Some(v) => v
-    case None => null
-    case v: Any => v
+  private def recordToSeq(record: Record): Seq[(String, Value)] =
+    record.fields.asScala.map(p => p.key -> p.value)
+
+  private def convertParams(params: Map[String, Any]): JMap[String, Object] =
+    params.mapValues(toNeoType).asJava
+
+  private def toNeoType(value: Any): AnyRef = value match {
+    case s: Seq[_]    => s.map(toNeoType).asJava
+    case s: Set[_]    => s.map(toNeoType).asJava
+    case m: Map[_, _] => m.mapValues(toNeoType).asJava
+    case o: Option[_] => o.map(toNeoType).orNull
+    case s: String    => s
+    case c: Char      => c: java.lang.Character
+    case b: Boolean   => b: java.lang.Boolean
+    case i: Int       => i: java.lang.Integer
+    case l: Long      => l: java.lang.Long
+    case f: Float     => f: java.lang.Float
+    case d: Double    => d: java.lang.Double
```

Comment:
I think there still should be `case v: Any => v` e.g. for LocalDate, LocalTime, etc

Reply:
The problem is that now, it is a **Function** from `Any` to `AnyRef`, so that won't compile.
We can do `case v: Any => v.asInstanceOf[AnyRef]` it is a _safe_ cast, since all primitive types are already handled and we could even remove them, since they will be automatically boxed. But it is still a `asInstanceOf`.

I forgot that you added some custom types, I think it would be better to add all possibilities here.
Which ones should they be, all that had an instance of **ValueMapper**?

The only problem is, does an user can create a custom type?
If so, the unsafe cast should be made at the end anyways.

Reply:
This part is responsible for passing parameters (not result mapping) and it's currently unsafe (passed as `params: Map[String, Any]`).
I didn't come up with a good way to do it more type-safely - as I remember, **doobie** doesn't support it too. My initial idea was to use shapeless records but it's no intuitive and doesn't look concise.

I you have good idea on how to implement it I'd be happy to discuss.

Reply:
Yeah, yeah. It is already ugly. As I said the `asInstanceOf[AnyRef]` is _safe_ in the sense it will always be correct. So, if we can not know by had which types we could have, I think that would be the best option.

Anyways, going to sleep.
Will see your answer tomorrow, and will push the final changes ASAP.
BTW, little spoiler, I realized we are missing a `Set` **ValueMapper** and a `set` method on queries, will add them too 😉  👍

Reply:
`asInstanceOf[AnyRef]` works for me. 
Oh yeah, seems I missed `Set`

Reply:
@BalmungSan would you like me to make this change?

Reply:
@dimafeng I should have it done on a couple of hours or so.
Tomorrow on the worst case. :+1:

Reply:
Sorry, didn't mean to hurry you, just wanted to offer some help 👍